src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph_RemoveProjectReference.cs

```diff
@@ -0,0 +1,130 @@
+        }
+
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? ComputeNewReverseReferencesMapForRemovedProjectReference(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? existingReverseReferencesMap,
+            ProjectId projectId,
+            ProjectId referencedProjectId)
+        {
+            if (existingReverseReferencesMap is null)
+                return null;
+
+            if (existingReverseReferencesMap.TryGetValue(referencedProjectId, out var referencingProjects))
+            {
+                return existingReverseReferencesMap.SetItem(referencedProjectId, referencingProjects.Remove(projectId));
+            }
+
+            return existingReverseReferencesMap;
+        }
+
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewTransitiveReferencesMapForRemovedProjectReference(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingTransitiveReferencesMap,
+            ProjectId projectId,
+            ProjectId referencedProjectId)
+        {
+            var builder = existingTransitiveReferencesMap.ToBuilder();
+
+            // Invalidate the transitive references from every project referencing the changed project (transitively)
+            foreach (var (project, references) in existingTransitiveReferencesMap)
+            {
+                if (references.Contains(projectId))
+                {
+                    builder.Remove(project);
+                }
+            }
+
+            // Invalidate the transitive references from the changed project
+            builder.Remove(projectId);
+
+            return builder.ToImmutable();
+        }
+
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseTransitiveReferencesMapForRemovedProjectReference(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseTransitiveReferencesMap,
+            ProjectId projectId,
+            ProjectId referencedProjectId)
+        {
+            var builder = existingReverseTransitiveReferencesMap.ToBuilder();
+
+            // Invalidate the transitive reverse references from every project previously referenced by the original
+            // project (transitively), except for cases where the removed project reference could not impact the result.
+            foreach (var (project, references) in existingReverseTransitiveReferencesMap)
+            {
+                if (!references.Contains(referencedProjectId))
+                {
+                    // If projectId references project, it isn't through referencedProjectId so the change doesn't
+                    // impact the dependency graph.
+                    //
+                    // Suppose we start with the following graph, and are removing the project reference A->B:
+                    //
+                    //   A -> B -> C
+                    //   A -> D
```

Comment:
Slightly fancier ASCII art -- at first I thought this was two independent chains.

```
                    //   A -> B -> C
                    //     \
                    //      > D
```